crates/edr_solidity/src/source_map.rs

```diff
@@ -154,52 +157,53 @@ pub fn decode_instructions(
         let source_map = &source_maps[instructions.len()];
 
         let pc = bytes_index;
-        let opcode = Opcode::from_repr(bytecode[pc]).expect("Invalid opcode");
+        let opcode = OpCode::new(bytecode[pc]).expect("Invalid opcode");
 
-        let push_data = if is_push(opcode) {
-            let length = opcode.push_len();
-            let push_data = &bytecode[(bytes_index + 1)..(bytes_index + 1 + (length as usize))];
+        let push_data = if opcode.is_push() {
+            let push_data = &bytecode[bytes_index..][..1 + opcode.info().immediate_size() as usize];
 
-            Some(Buffer::from(push_data))
+            Some(push_data.to_vec())
         } else {
             None
         };
 
-        let jump_type = if is_jump(opcode) && source_map.jump_type == JumpType::NOT_JUMP {
-            JumpType::INTERNAL_JUMP
-        } else {
-            source_map.jump_type
+        let jump_type = match (opcode, source_map.jump_type) {
+            (OpCode::JUMP | OpCode::JUMPI, JumpType::NotJump) => JumpType::InternalJump,
+            _ => source_map.jump_type,
         };
 
         let location = if source_map.location.file == -1 {
             None
         } else {
-            match file_id_to_source_file.get(&(source_map.location.file as u32)) {
-                Some(file) => {
-                    let location = SourceLocation::new(
-                        file.clone(),
+            build_model
+                .file_id_to_source_file
+                .get(&(source_map.location.file as u32))
+                .map(|_| {
+                    Rc::new(SourceLocation::new(
+                        build_model.file_id_to_source_file.clone(),
+                        source_map.location.file as u32,
                         source_map.location.offset as u32,
                         source_map.location.length as u32,
-                    )
-                    .into_instance(env)?;
-
-                    Some(ClassInstanceRef::from_obj(location, env)?)
-                }
-                None => None,
-            }
+                    ))
+                })
         };
 
-        let instruction =
-            Instruction::new(bytes_index as u32, opcode, jump_type, push_data, location)?;
+        let instruction = Instruction {
+            pc: bytes_index as u32,
+            opcode,
+            jump_type,
+            push_data,
+            location,
+        };
 
         instructions.push(instruction);
 
-        bytes_index += opcode.len() as usize;
+        bytes_index += 1 + opcode.info().immediate_size() as usize;
```

Comment:
Why `1 +`?

Reply:
If I understand `immediate_size` correctly, it's the size of the data that follows the opcode, so for PUSH1 that's 1. The original code used a `len` fn that returned `(1 + push_data.len())` for PUSHx or 1 otherwise, so I understood that this moves the index past the current opcode (hence the `1`) and the immediate size following that.